dagger/run_tests.py

```diff
@@ -0,0 +1,55 @@
+# referenced with modifications from:
+# https://docs.dagger.io/sdk/python/628797/get-started
+"""Run tests for multiple Python versions concurrently."""
+
+import sys
+
+import anyio
+
+import dagger
+
+
+async def test(versions):
+
+    async with dagger.Connection(dagger.Config(log_output=sys.stderr)) as client:
+        # get reference to the local project
+        src = client.host().directory(".")
+
+        async def test_version(version: str):
+            python = (
+                client.container()
+                .from_(f"python:{version}-slim-buster")
+                # mount cloned repository into image
+                .with_mounted_directory("/demo-github-actions", src)
+                # set current working directory for next commands
+                .with_workdir("/demo-github-actions")
+                # install test dependencies
+                .with_exec(["pip", "install", "poetry==1.4"])
+                .with_exec(["poetry", "install"])
+                # run tests
+                .with_exec(["poetry", "run", "python", "-m", "pytest"])
+            )
+
+            print(f"Starting tests for Python {version}")
+
+            # execute
+            await python.exit_code()
+
+            print(f"Tests for Python {version} succeeded!")
+
+        # when this block exits, all tasks will be awaited (i.e., executed)
+        async with anyio.create_task_group() as tg:
+            for version in versions:
+                tg.start_soon(test_version, version)
+
+    print("All tasks have finished")
+
+
+if __name__ == "__main__":
+    if len(sys.argv) > 1:
+        py_versions = sys.argv[1:]
+    else:
+        py_versions = ["3.10", "3.11"]
+
+    anyio.run(test, py_versions)
```

Comment:
Personally, I'd prefer to see this code throw an error if it doesn't get the arguments it expects rather than using an undocumented default set of versions, just to minimize surprises for the end-user. If it's important for it to run using default values when none are given, it'd be nice to see a message about that prior to it running.

Reply:
Thank you, great point! The defaults here allow for fewer keystrokes when using this as a development tool and further specify which Python versions are expected to be tested by default (perhaps as an indication of support within the code). I found myself wanting to somehow reference the `pyproject.toml` file somehow but knowing this might introduce some complexity into an otherwise simple implementation (the reference as-is would need to somehow know there's no Python `3.18` yet). Maybe using specific versions would be enough here. I'll see if there's a way to make this more elegant by referencing a single environment file by default.

Reply:
It makes sense that you'd like people to be able to rapidly run tests, and I do generally like the idea of having the versions of Python you're targeting in the project metadata. Since this just an example I think it's fine to either leave this as-is, or to write a simple message that it's testing those two versions since none were specified, assuming that doesn't make the test output messy or difficult to parse.

I agree that parsing the `pyproject.toml` would complicate things too much for this simple example, so using hardcoded versions is IMHO fine.

Reply:
I've been meaning to figure this out in some other areas too, the duplicative specification gets to be tricky and can be a development time drain. I'll still give metadata parsing a try here and add a commit if it's reasonable, following up with you for feedback/thoughts. No matter what, I'll also add a message stating the default functionality, as you mentioned.

Reply:
I found a way to make this work but wasn't happy about how complex it became and the reliance on `eval` (or `ast.literal_eval`). I ended up only adding the message as you suggested, especially to keep the topic focused.

For reference, this was what I found might work for something like this in the future. Perhaps this can be improved over time. Python 3.11 includes `toml` support natively so maybe this will become more commonplace over time.
```python
   try:
        import tomllib
    except ModuleNotFoundError:
        import tomli as tomllib

    with open("pyproject.toml", "rb") as f:
        python_versions = tomllib.load(f)["tool"]["poetry"]["dependencies"]["python"]

    # loop on potential python versions from 3.0 - 4.0
    # note: divide by 100 to get potential float number
    for x in [x / 100 for x in range(300, 400, 1)]:
        # eval python version using poetry toml specification
        # note: assumes min and max range are specified within the python version string
        # example: '>=3.10,<3.12'
        if eval("x " + " and x ".join(python_versions.split(","))):
            # format version string to retain zeros
            print("%.02f" % x)
```

Reply:
Agreed about it being too complicated for now; we can discuss it later if you like. I also definitely agree about `eval` being problematic; it might be worth trying to pull the set of available Python versions, then try to identify which actual Python versions match the version range expressions. I should caution you that comparing semantic versions is tricky; you might consider using the [packaging](https://pypi.org/project/packaging/) module to make that easier.

Reply:
Thanks - would love to discuss this later, appreciate the `packaging` reference!